File: junit-examples/junit-jupiter-starter-maven/src/test/java/com/example/project/AssertionDensityCalculator.java
package com.example.project;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.regex.Pattern;

/**
 * Helper executed by exec-maven-plugin to compute assertion density
 * (number of assert* calls per @Test / @ParameterizedTest).
 */
public final class AssertionDensityCalculator {

    private static final Pattern TEST_ANNOTATION =
            Pattern.compile("@Test\\b|@ParameterizedTest\\b");

    private static final Pattern ASSERTION =
            Pattern.compile("\\bassert(All|Equals|NotEquals|Null|NotNull|True|False|Throws|DoesNotThrow|IterableEquals)\\b");

    public static void main(String[] args) throws IOException {
        String tagFilter = System.getProperty("junit.jupiter.tags");
        Path testRoot = Paths.get("src", "test", "java");
        if (!Files.exists(testRoot)) {
            System.out.println("No test sources found at " + testRoot.toAbsolutePath());
            return;
        }

        final int[] tests = {0};
        final int[] assertions = {0};

        Files.walk(testRoot)
                .filter(p -> p.toString().endsWith(".java"))
                .forEach(p -> {
                    try {
                        String text = Files.readString(p);
                        if (tagFilter != null && !tagFilter.isBlank()) {
                            String tagToken = "@Tag(\"" + tagFilter.trim() + "\")";
                            if (!text.contains(tagToken)) {
                                return;
                            }
                        }
                        tests[0] += countMatches(TEST_ANNOTATION, text);
                        assertions[0] += countMatches(ASSERTION, text);
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                });

        if (tests[0] == 0) {
            System.out.println("No @Test/@ParameterizedTest methods found.");
        } else {
            double density = assertions[0] / (double) tests[0];
            System.out.printf(
                    "Assertion density: %d assertions / %d tests = %.3f%n",
                    assertions[0], tests[0], density
            );
        }
    }

    private static int countMatches(Pattern pattern, String text) {
        int count = 0;
        var m = pattern.matcher(text);
        while (m.find()) {
            count++;
        }
        return count;
    }

    private AssertionDensityCalculator() {
        // utility class
    }
}

File: junit-examples/junit-jupiter-starter-maven-kotlin/src/test/kotlin/com/example/project/AssertionDensityCalculator.kt
package com.example.project

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Computes a simple assertion density metric: number of assert* calls per @Test/@ParameterizedTest.
 * Invoked via exec-maven-plugin.
 */
fun main() {
    val tagFilter = System.getProperty("junit.jupiter.tags")?.trim()?.takeIf { it.isNotEmpty() }
    val roots = listOf(
        Paths.get("src", "test", "kotlin"),
        Paths.get("src", "test", "java")
    )

    val existingRoots = roots.filter { Files.exists(it) }
    if (existingRoots.isEmpty()) {
        println("No test sources found in src/test/kotlin or src/test/java")
        return
    }

    val testPattern = Regex("@Test\\b|@ParameterizedTest\\b")
    val assertPattern = Regex("\\bassert(All|Equals|NotEquals|Null|NotNull|True|False|Throws|DoesNotThrow|IterableEquals)\\b")

    var testCount = 0
    var assertionCount = 0

    existingRoots.forEach { root ->
        Files.walk(root).use { paths ->
            paths.filter { Files.isRegularFile(it) && (it.toString().endsWith(".kt") || it.toString().endsWith(".java")) }
                .forEach { path ->
                    val text = Files.readString(path)
                    if (tagFilter != null && !text.contains("@Tag(\"$tagFilter\")")) {
                        return@forEach
                    }
                    testCount += testPattern.findAll(text).count()
                    assertionCount += assertPattern.findAll(text).count()
                }
        }
    }

    if (testCount == 0) {
        println("No @Test/@ParameterizedTest methods found.")
    } else {
        val density = assertionCount.toDouble() / testCount.toDouble()
        val tagLabel = tagFilter?.let { " (tag=$it)" } ?: ""
        println("Assertion density$tagLabel: $assertionCount assertions / $testCount tests = %.3f".format(density))
    }
}
